src/FastConv/Convolver.cpp

```diff
@@ -0,0 +1,323 @@
+    m_H_real = new float* [m_IRNumBlock];
+    m_H_imag = new float* [m_IRNumBlock];
+
+    // initialize buffer for temporary results
+    m_buffer = new CRingBuffer<float>((m_IRNumBlock + 1) * blockLength);
+
+    // initialize the temp spaces
+    aReal = new float[blockLength + 1]; memset(aReal, 0, sizeof(float) * (blockLength + 1));
+    cReal = new float[blockLength + 1]; memset(cReal, 0, sizeof(float) * (blockLength + 1));
+    aImag = new float[blockLength + 1]; memset(aImag, 0, sizeof(float) * (blockLength + 1));
+    cImag = new float[blockLength + 1]; memset(cImag, 0, sizeof(float) * (blockLength + 1));
+    temp  = new float[blockLength + 1]; memset(temp,  0, sizeof(float) * (blockLength + 1));
+    aReal_bReal = new float[blockLength + 1]; memset(aReal_bReal,  0, sizeof(float) * (blockLength + 1));
+    aImag_bImag = new float[blockLength + 1]; memset(aImag_bImag,  0, sizeof(float) * (blockLength + 1));
+    iFFTTemp = new float[doubleBlockLength]; memset(iFFTTemp,  0, sizeof(float) * (doubleBlockLength));
+
+    // allocate memory for frequency domain
+    m_X = new CFft::complex_t[doubleBlockLength];
+    m_X_origin = new CFft::complex_t[doubleBlockLength];
+
+    // temporary variable for blocked IR freq representatie for blocked IR freq representation
+    auto m_H = new CFft::complex_t[doubleBlockLength];
+
+    // pre-calculate the spectrogram of the impulse response
+    for (int i = 0; i < m_IRNumBlock; ++i) {
+        m_H_real[i] = new float[blockLength + 1]; memset(m_H_real[i], 0, sizeof(float) * (blockLength + 1));
+        m_H_imag[i] = new float[blockLength + 1]; memset(m_H_imag[i], 0, sizeof(float) * (blockLength + 1));
+
+        memset(iFFTTemp, 0, sizeof(float) * doubleBlockLength);
+        memcpy(iFFTTemp, impulseResponse + (i * blockLength), sizeof(float) * std::min(blockLength, m_originIRLengthInSample - i * blockLength));
+
+        m_pCFft->doFft(m_H, iFFTTemp);
+        m_pCFft->splitRealImag(m_H_real[i], m_H_imag[i], m_H);
+    }
+
+    delete[] m_H;
+
+    return Error_t::kNoError;
+}
+
+Error_t UniformlyPartitionedFFTConvolver::reset() {
+    for (int i = 0; i < m_IRNumBlock; ++i) {
+        delete[] m_H_real[i];
+        delete[] m_H_imag[i];
+    }
+    delete[] m_H_real;
+    delete[] m_H_imag;
+    delete[] m_X;
+
+    delete[] aReal;
+    delete[] cReal;
+    delete[] aImag;
+    delete[] cImag;
+    delete[] aReal_bReal;
+    delete[] aImag_bImag;
+
+    delete[] iFFTTemp;
+    delete[] temp;
+
+    delete m_buffer;
+
+    m_pCFft->resetInstance();
+    CFft::destroyInstance(m_pCFft);
+    m_pCFft = nullptr;
+
+    return Error_t::kNoError;
+}
+
+Error_t UniformlyPartitionedFFTConvolver::process(float* output, const float* input, int bufferLength) {
+    int accuLength = 0;
+    int curBufferLength;
+    while (bufferLength) {
+        curBufferLength = std::min(bufferLength, m_blockLength);
+        __processOneBlock(output + accuLength, input + accuLength, curBufferLength);
```

Comment:
I'm a bit suspicious of how this goes ahead and takes an FFT immediately, even if the next block hasn't been completely filled. (This is related to my other comments about the apparent missing latency and the `varyingBlockSize` test.)

Reply:
Oh, this assumes that `output` and `input` has at least `bufferLength` samples available, and this function deals with case when `bufferLength > m_blockLength`. When `bufferLength <= m_blockLength`, it will be processed in `__processOneBlock`. I think this design makes convolver more robust, and also omits low level details of the computational procedure.